_posts/2020-12-22-aws-antipatterns.md

```diff
@@ -0,0 +1,198 @@
+
+![Bills can creep up on you unless you keep an eye on them](/img/aws-antipatterns/trusted_advisor.png)
+
+It's possible to do AWS in a very restricted and controlled manner, by going through slow and expensive ticketing for each resource, but this is not the antipattern we're talking about here. It would be worse than an antipattern because it would kill all the benefits mentioned here. But most companies use a model where the DevOps-teams have a lot of access within the account given to them, and traditional IT administration is only creating and monitoring the accounts. This model gives the benefits of speed and cost-effectiveness.
+
+When a team starts building stuff within AWS, in an agile manner, they pretty soon discover that given enough power, they can create and scale resources pretty much immediately, even dynamically. This supports agile operations very well. Many of the resources are extremely cheap, and you get a lot of extra options that are ridiculously cheap to turn on and enjoy immediate benefits. Things like setting up the network to your needs, using encryption-at-rest or in-transit, getting storage without needing to worry about CPUs, licenses, or memory. Getting redundancy and disaster recovery by just clicking a few buttons. And it's all very cheap, compared to setting up the servers to do all that yourself. And this is where the trouble starts.
+
+You see, AWS has more than 175 products and services, and more are coming each year. Each has a different pricing model. Due to being able to immediately toggle on new services, which they have varying levels of experience, it's relatively easy to turn on something that would cost a lot. Here are some examples:
+
+- I've once seen a badly written SQL query that was being run in Athena every hour by a timed lambda. It was artfully missing all nicely generated indexes, and going through terabytes of data in the bucket, all of it. It was generating tens of thousands of euros of billing before it was discovered.
+- Creating virtual servers, it's possible to choose their type and capacity, and some of them can be very expensive if they are left running. Machine learning projects might often run these servers, but are typically not supposed to run them 24/7, only on demand.
+- Sagemaker developer endpoints can also rack up some bill quite fast, it's easy to create them and use them, but if you forget to shut them down, euros will start accumulating rapidly
+- There are often few ways to model things, and the cost may vary quite a lot. Different patterns depending on your case.
+- If you're not seeing how much things cost, you might be enticed to create a new feature for the product, based on some new service just because you can. Perhaps it's something you wouldn't have even considered before.
+
+This can often be a problem because I've seen that billing is often abstracted, denied, and hidden from the development team. So they can freely 
```

Comment:
"So they can freely", how does this continue?

Reply:
Ah, good find! This is not finished, I will fix this,.